File: _includes/code/csharp/ManageObjectsUpdateTest.cs
using Xunit;
using Weaviate.Client;
using Weaviate.Client.Models;
using System;
using System.Threading.Tasks;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace WeaviateProject.Tests;

public class ManageObjectsUpdateTest : IAsyncLifetime
{
    private static readonly WeaviateClient client;

    // Static constructor for one-time setup
    static ManageObjectsUpdateTest()
    {
        // START INSTANTIATION-COMMON
        // Note: The C# client doesn't support setting headers like 'X-OpenAI-Api-Key' via the constructor.
        // This must be configured in Weaviate's environment variables.
        client = new WeaviateClient(new ClientConfiguration { RestAddress = "localhost", RestPort = 8080 });
        // END INSTANTIATION-COMMON
    }

    // Runs once before any tests in the class (like @BeforeAll)
    public async Task InitializeAsync()
    {
        // Simulate weaviate-datasets and set up collections
        if (await client.Collections.Exists("WineReviewNV"))
        {
            await client.Collections.Delete("WineReviewNV");
        }
        await client.Collections.Create(new Collection
        {
            Name = "WineReviewNV",
            Properties =
            [
                Property.Text("review_body", description: "Review body"),
                Property.Text("title", description: "Name of the wine"),
                Property.Text("country", description: "Originating country")
            ],
            VectorConfig = new[]
            {
                new VectorConfig("title", new Vectorizer.Text2VecContextionary()),
                new VectorConfig("review_body", new Vectorizer.Text2VecContextionary()),
                new VectorConfig(
                "title_country",
                new Vectorizer.Text2VecContextionary { SourceProperties = ["title", "country"] }
                )
            }
        });

        // highlight-start
        // ===== Add three mock objects to the WineReviewNV collection =====
        var reviews = client.Collections.Use<object>("WineReviewNV");
        await reviews.Data.InsertMany(new[]
        {
            new { title = "Mock Wine A", review_body = "A fine mock vintage.", country = "Mocktugal" },
            new { title = "Mock Wine B", review_body = "Notes of mockberry.", country = "Mockstralia" },
            new { title = "Mock Wine C", review_body = "Pairs well with mock turtle soup.", country = "Republic of Mockdova" }
        });
        // highlight-end

        // START Define the class
        if (await client.Collections.Exists("JeopardyQuestion"))
        {
            await client.Collections.Delete("JeopardyQuestion");
        }
        await client.Collections.Create(new Collection
        {
            Name = "JeopardyQuestion",
            Description = "A Jeopardy! question",
            Properties =
            [
                Property.Text("question", description: "The question"),
                Property.Text("answer", description: "The answer"),
                Property.Number("points", description: "The points the question is worth")
            ],
            VectorConfig = new VectorConfig("default", new Vectorizer.Text2VecContextionary())
        });
        // END Define the class
    }

    // Runs once after all tests in the class (like @AfterAll)
    public async Task DisposeAsync()
    {
        await client.Collections.Delete("WineReviewNV");
        await client.Collections.Delete("JeopardyQuestion");
    }

    // START DelProps
    private static async Task DelProps(WeaviateClient client, Guid uuidToUpdate, string collectionName,
        IEnumerable<string> propNames)
    {
        var collection = client.Collections.Use<object>(collectionName);

        // fetch the object to update
        var objectData = await collection.Query.FetchObjectByID(uuidToUpdate);
        if (objectData?.Properties is not IDictionary<string, object> propertiesToUpdate)
        {
            return;
        }

        // remove unwanted properties
        foreach (var propName in propNames)
        {
            propertiesToUpdate.Remove(propName);
        }

        // replace the properties
        await collection.Data.Replace(uuidToUpdate, propertiesToUpdate);
    }
    // END DelProps

    [Fact]
    public async Task TestUpdateAndReplaceFlow()
    {
        var jeopardy = client.Collections.Use<object>("JeopardyQuestion");

        var uuid = await jeopardy.Data.Insert(new
        {
            question = "Test question",
            answer = "Test answer",
            points = -1
        });

        // START UpdateProps
        await jeopardy.Data.Replace(uuid,
        // highlight-start
         data: new { points = 100 }
        // highlight-end
        );
        // END UpdateProps

        var result1 = await jeopardy.Query.FetchObjectByID(uuid);
        Assert.NotNull(result1);
        var props1 = result1.Properties as IDictionary<string, object>;
        Assert.Equal(100d, props1["points"]);


        var vector = Enumerable.Repeat(0.12345f, 300).ToArray();

        // TODO[g-despot]  Not implemented
        // START UpdateVector        
        // Coming soon
        // END UpdateVector
        // await jeopardy.Data.Update(uuid,
        //     properties: new { points = 100 },
        //     // highlight-start
        //     vector: vector
        // // highlight-end
        // );

        // var result2 = await jeopardy.Query.FetchObjectByID(uuid, returnMetadata: MetadataOptions.Vector);
        // Assert.NotNull(result2);
        // Assert.Equal(300, result2.Vectors["default"].Dimensions);


        // TODO[g-despot]  Not implemented
        // START UpdateNamedVector
        // Coming soon
        // END UpdateNamedVector

        var reviews = client.Collections.Use<object>("WineReviewNV");
        var reviewResponse = await reviews.Query.FetchObjects(limit: 1);
        var reviewUuid = reviewResponse.Objects.First().ID.Value;

        var titleVector = Enumerable.Repeat(0.12345f, 300).ToArray();
        var reviewBodyVector = Enumerable.Repeat(0.23456f, 300).ToArray();
        var titleCountryVector = Enumerable.Repeat(0.34567f, 300).ToArray();

        // await reviews.Data.Update(reviewUuid,
        //     data: new
        //     {
        //         title = "A delicious wine",
        //         review_body = "This mystery wine is a delight to the senses.",
        //         country = "Mordor"
        //     },
        //     // highlight-start
        //     vectors: new Dictionary<string, float[]>
        //     {
        //         { "title", titleVector },
        //         { "review_body", reviewBodyVector },
        //         { "title_country", titleCountryVector }
        //     }
        //     // highlight-end
        // );


        // START Replace
        // highlight-start
        await jeopardy.Data.Replace(
            // highlight-end
            uuid,
            data: new { answer = "Replaced" }
        // The other properties will be deleted
        );
        // END Replace

        var result3 = await jeopardy.Query.FetchObjectByID(uuid);
        Assert.NotNull(result3);
        var props3 = result3.Properties as IDictionary<string, object>;
        Assert.Equal("Replaced", props3["answer"].ToString());
        Assert.DoesNotContain("question", props3.Keys);

        // START DelProps

        await DelProps(client, uuid, "JeopardyQuestion", new[] { "answer" });
        // END DelProps

        var result4 = await jeopardy.Query.FetchObjectByID(uuid);
        Assert.NotNull(result4);
        var props4 = result4.Properties as IDictionary<string, object>;
        Assert.False(props4.ContainsKey("answer"));
    }
}

File: _includes/code/java-v6/src/test/java/ManageObjectsUpdateTest.java
import io.weaviate.client6.v1.api.WeaviateClient;
import io.weaviate.client6.v1.api.collections.CollectionHandle;
import io.weaviate.client6.v1.api.collections.Property;
import io.weaviate.client6.v1.api.collections.VectorConfig;
import io.weaviate.client6.v1.api.collections.Vectors;
import io.weaviate.client6.v1.api.collections.WeaviateObject;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

class ManageObjectsUpdateTest {

  private static WeaviateClient client;

  @BeforeAll
  public static void beforeAll() throws IOException {
    // START INSTANTIATION-COMMON
    // Instantiate the client with the OpenAI API key
    String openaiApiKey = System.getenv("OPENAI_API_KEY");
    client = WeaviateClient
        .connectToLocal(config -> config.setHeaders(Map.of("X-OpenAI-Api-Key", openaiApiKey)));
    // END INSTANTIATION-COMMON

    // Simulate weaviate-datasets and set up collections
    if (client.collections.exists("WineReviewNV")) {
      client.collections.delete("WineReviewNV");
    }
    client.collections.create("WineReviewNV",
        col -> col
            .properties(Property.text("review_body", p -> p.description("Review body")),
                Property.text("title", p -> p.description("Name of the wine")),
                Property.text("country", p -> p.description("Originating country")))
            .vectorConfig(VectorConfig.text2vecTransformers("title"),
                VectorConfig.text2vecTransformers("review_body"),
                VectorConfig.text2vecTransformers("title_country",
                    vc -> vc.sourceProperties("title", "country"))));

    // highlight-start
    // ===== Add three mock objects to the WineReviewNV collection =====
    var reviews = client.collections.use("WineReviewNV");
    reviews.data.insertMany(
        Map.of("title", "Mock Wine A", "review_body", "A fine mock vintage.", "country",
            "Mocktugal"),
        Map.of("title", "Mock Wine B", "review_body", "Notes of mockberry.", "country",
            "Mockstralia"),
        Map.of("title", "Mock Wine C", "review_body", "Pairs well with mock turtle soup.",
            "country", "Republic of Mockdova"));
    // highlight-end

    // START Define the class
    if (client.collections.exists("JeopardyQuestion")) {
      client.collections.delete("JeopardyQuestion");
    }
    client.collections.create("JeopardyQuestion",
        col -> col.description("A Jeopardy! question")
            .properties(Property.text("question", p -> p.description("The question")),
                Property.text("answer", p -> p.description("The answer")),
                Property.number("points", p -> p.description("The points the question is worth")))
            .vectorConfig(VectorConfig.text2vecTransformers()));
    // END Define the class
  }

  @AfterAll
  public static void afterAll() throws Exception {
    client.collections.delete("WineReviewNV");
    client.collections.delete("JeopardyQuestion");
    client.close();
  }

  // START DelProps
  private static void delProps(WeaviateClient client, String uuidToUpdate, String collectionName,
      List<String> propNames) throws IOException {
    CollectionHandle<Map<String, Object>> collection = client.collections.use(collectionName);

    // fetch the object to update
    Optional<WeaviateObject<Map<String, Object>>> objectDataOpt =
        collection.query.fetchObjectById(uuidToUpdate);
    if (objectDataOpt.isEmpty()) {
      return;
    }
    Map<String, Object> propertiesToUpdate = new HashMap<>(objectDataOpt.get().properties());

    // remove unwanted properties
    for (String propName : propNames) {
      propertiesToUpdate.remove(propName);
    }

    // replace the properties
    collection.data.replace(uuidToUpdate, r -> r.properties(propertiesToUpdate));
  }
  // END DelProps

  @Test
  void testUpdateAndReplaceFlow() throws IOException {
    CollectionHandle<Map<String, Object>> jeopardy = client.collections.use("JeopardyQuestion");

    String uuid = jeopardy.data
        .insert(Map.of("question", "Test question", "answer", "Test answer", "points", -1.0 // JSON numbers are doubles
        )).uuid();

    // START UpdateProps
    jeopardy.data.update(uuid,
        // highlight-start
        u -> u.properties(Map.of("points", 100.0))
    // highlight-end
    );
    // END UpdateProps

    Optional<WeaviateObject<Map<String, Object>>> result1 =
        jeopardy.query.fetchObjectById(uuid);
    assertThat(result1).isPresent();
    assertThat(result1.get().properties().get("points")).isEqualTo(100.0);

    // START UpdateVector
    float[] vector = new float[384];
    Arrays.fill(vector, 0.12345f);

    jeopardy.data.update(uuid, u -> u.properties(Map.of("points", 100.0))
        // highlight-start
        .vectors(Vectors.of(vector))
    // highlight-end
    );
    // END UpdateVector

    Optional<WeaviateObject<Map<String, Object>>> result2 =
        jeopardy.query.fetchObjectById(uuid, q -> q.includeVector());
    assertThat(result2).isPresent();
    assertThat(result2.get().vectors().getSingle("default")).hasSize(384);

    // START UpdateNamedVector
    CollectionHandle<Map<String, Object>> reviews = client.collections.use("WineReviewNV");
    String reviewUuid = reviews.query.fetchObjects(q -> q.limit(3)).objects().get(0).uuid();
    float[] titleVector = new float[384];
    float[] reviewBodyVector = new float[384];
    float[] titleCountryVector = new float[384];
    Arrays.fill(titleVector, 0.12345f);
    Arrays.fill(reviewBodyVector, 0.12345f);
    Arrays.fill(titleCountryVector, 0.12345f);

    reviews.data.update(reviewUuid,
        u -> u
            .properties(Map.of("title", "A delicious wine", "review_body",
                "This mystery wine is a delight to the senses.", "country", "Mordor"))
            // highlight-start
            .vectors(Vectors.of("title", titleVector), Vectors.of("review_body", reviewBodyVector),
                Vectors.of("title_country", titleCountryVector))
    // highlight-end
    );
    // END UpdateNamedVector

    // START Replace
    // highlight-start
    jeopardy.data.replace(
        // highlight-end
        uuid, r -> r.properties(Map.of("answer", "Replaced"
        // The other properties will be deleted
        )));
    // END Replace

    Optional<WeaviateObject<Map<String, Object>>> result3 =
        jeopardy.query.fetchObjectById(uuid);
    assertThat(result3).isPresent();
    assertThat(result3.get().properties().get("answer")).isEqualTo("Replaced");

    // START DelProps

    delProps(client, uuid, "JeopardyQuestion", List.of("answer"));
    // END DelProps

    Optional<WeaviateObject<Map<String, Object>>> result4 =
        jeopardy.query.fetchObjectById(uuid);
    assertThat(result4).isPresent();
    assertThat(result4.get().properties().get("answer")).isNull();
  }
}
